File: backend/narrow_edit_idempotency.py
"""Bounded in-process receipts for Ren's narrow prompt and mask mutations.

The browser remains the durable authority for a ComfyUI page session.  This
ledger only closes duplicate-call races inside one MCP subprocess and keeps no
raw request payloads (notably, no prompt text).  A caller-provided operation ID
is bound to a canonical request digest; reusing it for different arguments is
always an error.
"""

from __future__ import annotations

import hashlib
import math
import re
import struct
import threading
import time
from copy import deepcopy
from dataclasses import dataclass
from typing import Any, Callable, Literal, Mapping

NARROW_EDIT_OPERATION_SCHEMA = "fl-mcp.narrow-edit-operation.v1"
NARROW_EDIT_OPERATION_ID_PATTERN = re.compile(
    r"^[A-Za-z0-9][A-Za-z0-9._:-]{7,127}$"
)
NARROW_EDIT_REQUEST_HASH_PATTERN = re.compile(r"^[0-9a-f]{64}$")


class NarrowEditIdempotencyError(ValueError):
    """A bounded, classified idempotency failure safe for user diagnostics."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


def validate_narrow_operation_id(operation_id: Any) -> str:
    """Validate one opaque, non-secret operation identifier."""

    if not isinstance(operation_id, str) or not NARROW_EDIT_OPERATION_ID_PATTERN.fullmatch(
        operation_id
    ):
        raise NarrowEditIdempotencyError(
            "narrow_edit_operation_id_invalid",
            "operation_id must be 8-128 characters using letters, digits, '.', '_', ':', or '-'.",
        )
    return operation_id


def _canonical_typed_bytes(value: Any) -> bytes:
    """Encode exact JSON facts consistently with the browser implementation."""

    item_type = type(value)
    if value is None:
        return b"n;"
    if item_type is bool:
        return b"b1;" if value else b"b0;"
    if item_type in {int, float}:
        try:
            number = float(value)
        except OverflowError as exc:
            raise NarrowEditIdempotencyError(
                "narrow_edit_payload_invalid",
                "Operation payload numbers must fit the browser Number domain.",
            ) from exc
        if not math.isfinite(number):
            raise NarrowEditIdempotencyError(
                "narrow_edit_payload_invalid",
                "Operation payload numbers must be finite.",
            )
        if number == 0:
            number = 0.0
        return b"d" + struct.pack(">d", number).hex().encode("ascii") + b";"
    if item_type is str:
        try:
            encoded = value.encode("utf-8")
        except UnicodeEncodeError as exc:
            raise NarrowEditIdempotencyError(
                "narrow_edit_payload_invalid",
                "Operation payload text must be valid UTF-8.",
            ) from exc
        return (
            b"s"
            + str(len(encoded)).encode("ascii")
            + b":"
            + encoded.hex().encode("ascii")
            + b";"
        )
    if item_type is list:
        return (
            b"a"
            + str(len(value)).encode("ascii")
            + b":["
            + b"".join(_canonical_typed_bytes(item) for item in value)
            + b"];"
        )
    if item_type is dict:
        if any(type(key) is not str for key in value):
            raise NarrowEditIdempotencyError(
                "narrow_edit_payload_invalid",
                "Operation payload object keys must be strings.",
            )
        encoded_items: list[bytes] = []
        for key in sorted(value):
            encoded_items.append(_canonical_typed_bytes(key))
            encoded_items.append(_canonical_typed_bytes(value[key]))
        return (
            b"o"
            + str(len(value)).encode("ascii")
            + b":{" + b"".join(encoded_items) + b"};"
        )
    raise NarrowEditIdempotencyError(
        "narrow_edit_payload_invalid",
        "Operation payloads must use the exact JSON data model.",
    )


def canonical_narrow_operation_hash(tool: str, payload: Mapping[str, Any]) -> str:
    """Hash one mutation request without retaining or logging its raw payload."""

    if not isinstance(tool, str) or not tool or len(tool) > 128:
        raise NarrowEditIdempotencyError(
            "narrow_edit_tool_invalid",
            "The narrow mutation tool name is invalid.",
        )
    if not isinstance(payload, Mapping):
        raise NarrowEditIdempotencyError(
            "narrow_edit_payload_invalid",
            "The narrow mutation payload must be an object.",
        )
    canonical_payload = {
        "schema": NARROW_EDIT_OPERATION_SCHEMA,
        "tool": tool,
        "payload": dict(payload),
    }
    return hashlib.sha256(_canonical_typed_bytes(canonical_payload)).hexdigest()


@dataclass(frozen=True)
class NarrowEditClaim:
    """One exact claim returned without exposing retained mutable state."""

    status: Literal["new", "pending", "completed"]
    session_id: str
    tool: str
    operation_id: str
    request_hash: str
    receipt: dict[str, Any] | None = None


@dataclass
class _LedgerEntry:
    session_id: str
    tool: str
    operation_id: str
    request_hash: str
    state: Literal["pending", "completed"]
    created_at: float
    expires_at: float
    receipt: dict[str, Any] | None = None


class NarrowEditOperationLedger:
    """Thread-safe TTL ledger that retains only digests and safe receipts."""

    def __init__(
        self,
        *,
        ttl_seconds: float = 30 * 60,
        max_entries: int = 256,
        clock: Callable[[], float] = time.monotonic,
    ) -> None:
        if not isinstance(ttl_seconds, (int, float)) or ttl_seconds <= 0:
            raise ValueError("ttl_seconds must be positive")
        if not isinstance(max_entries, int) or isinstance(max_entries, bool) or max_entries < 1:
            raise ValueError("max_entries must be a positive integer")
        self._ttl_seconds = float(ttl_seconds)
        self._max_entries = max_entries
        self._clock = clock
        self._entries: dict[tuple[str, str, str], _LedgerEntry] = {}
        self._lock = threading.RLock()

    @staticmethod
    def _key(session_id: str, tool: str, operation_id: str) -> tuple[str, str, str]:
        if not isinstance(session_id, str) or not session_id or len(session_id) > 256:
            raise NarrowEditIdempotencyError(
                "narrow_edit_session_invalid",
                "The browser session identity is invalid.",
            )
        if not isinstance(tool, str) or not tool or len(tool) > 128:
            raise NarrowEditIdempotencyError(
                "narrow_edit_tool_invalid",
                "The narrow mutation tool name is invalid.",
            )
        return session_id, tool, validate_narrow_operation_id(operation_id)

    @staticmethod
    def _request_hash(request_hash: Any) -> str:
        if not isinstance(request_hash, str) or not NARROW_EDIT_REQUEST_HASH_PATTERN.fullmatch(
            request_hash
        ):
            raise NarrowEditIdempotencyError(
                "narrow_edit_request_hash_invalid",
                "The narrow mutation request digest is invalid.",
            )
        return request_hash

    def _prune(self, now: float) -> None:
        for key, entry in list(self._entries.items()):
            if entry.expires_at <= now:
                del self._entries[key]

    def _make_room(self) -> None:
        if len(self._entries) < self._max_entries:
            return
        completed = [
            (key, entry)
            for key, entry in self._entries.items()
            if entry.state == "completed"
        ]
        if completed:
            oldest_key, _ = min(completed, key=lambda item: item[1].created_at)
            del self._entries[oldest_key]
            return
        raise NarrowEditIdempotencyError(
            "narrow_edit_ledger_capacity",
            "Too many narrow mutations are still pending; wait for one to finish before retrying.",
        )

    def claim(
        self,
        *,
        session_id: str,
        tool: str,
        operation_id: str,
        request_hash: str,
    ) -> NarrowEditClaim:
        """Claim or recover one exact operation without retaining its payload."""

        key = self._key(session_id, tool, operation_id)
        digest = self._request_hash(request_hash)
        with self._lock:
            now = self._clock()
            self._prune(now)
            existing = self._entries.get(key)
            if existing is not None:
                if existing.request_hash != digest:
                    raise NarrowEditIdempotencyError(
                        "narrow_edit_idempotency_conflict",
                        "operation_id was already used with different mutation arguments.",
                    )
                return NarrowEditClaim(
                    status=existing.state,
                    session_id=session_id,
                    tool=tool,
                    operation_id=operation_id,
                    request_hash=digest,
                    receipt=deepcopy(existing.receipt),
                )
            self._make_room()
            self._entries[key] = _LedgerEntry(
                session_id=session_id,
                tool=tool,
                operation_id=operation_id,
                request_hash=digest,
                state="pending",
                created_at=now,
                expires_at=now + self._ttl_seconds,
            )
            return NarrowEditClaim(
                status="new",
                session_id=session_id,
                tool=tool,
                operation_id=operation_id,
                request_hash=digest,
            )

    def complete(self, claim: NarrowEditClaim, receipt: Mapping[str, Any]) -> None:
        """Store a caller-sanitized receipt for exact replay."""

        if not isinstance(receipt, Mapping):
            raise NarrowEditIdempotencyError(
                "narrow_edit_receipt_invalid",
                "The narrow mutation receipt must be an object.",
            )
        key = self._key(claim.session_id, claim.tool, claim.operation_id)
        with self._lock:
            now = self._clock()
            self._prune(now)
            entry = self._entries.get(key)
            if entry is None or entry.request_hash != claim.request_hash:
                raise NarrowEditIdempotencyError(
                    "narrow_edit_claim_stale",
                    "The narrow mutation claim expired or changed before completion.",
                )
            entry.state = "completed"
            entry.receipt = deepcopy(dict(receipt))
            entry.expires_at = now + self._ttl_seconds

    def discard_pending(self, claim: NarrowEditClaim) -> bool:
        """Release one known pre-commit failure; completed receipts are immutable."""

        key = self._key(claim.session_id, claim.tool, claim.operation_id)
        with self._lock:
            entry = self._entries.get(key)
            if (
                entry is None
                or entry.request_hash != claim.request_hash
                or entry.state != "pending"
            ):
                return False
            del self._entries[key]
            return True

    def clear(self) -> None:
        with self._lock:
            self._entries.clear()

    def __len__(self) -> int:
        with self._lock:
            self._prune(self._clock())
            return len(self._entries)

File: web/js/narrow_edit_idempotency.js
/** Lost-response idempotency for Ren's narrow prompt and mask mutation lane. */

export const NARROW_EDIT_OPERATION_SCHEMA = "fl-mcp.narrow-edit-operation.v1";

const OPERATION_ID_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:-]{7,127}$/;
const SHA256_PATTERN = /^[0-9a-f]{64}$/;


function narrowEditError(code, message) {
    const error = new Error(message);
    error.code = code;
    return error;
}


export function validateNarrowOperationId(operationId) {
    if (typeof operationId !== "string" || !OPERATION_ID_PATTERN.test(operationId)) {
        throw narrowEditError(
            "narrow_edit_operation_id_invalid",
            "operation_id must be 8-128 characters using letters, digits, '.', '_', ':', or '-'.",
        );
    }
    return operationId;
}


function assertWellFormedString(value, path) {
    for (let index = 0; index < value.length; index += 1) {
        const code = value.charCodeAt(index);
        if (code >= 0xd800 && code <= 0xdbff) {
            const next = value.charCodeAt(index + 1);
            if (!(next >= 0xdc00 && next <= 0xdfff)) {
                throw narrowEditError(
                    "narrow_edit_payload_invalid",
                    `Operation payload text must be valid UTF-8 (at ${path}).`,
                );
            }
            index += 1;
        } else if (code >= 0xdc00 && code <= 0xdfff) {
            throw narrowEditError(
                "narrow_edit_payload_invalid",
                `Operation payload text must be valid UTF-8 (at ${path}).`,
            );
        }
    }
}


function utf8Hex(value, path) {
    assertWellFormedString(value, path);
    return [...new TextEncoder().encode(value)]
        .map(byte => byte.toString(16).padStart(2, "0"))
        .join("");
}


function compareUnicodeScalars(left, right) {
    const leftPoints = Array.from(left, character => character.codePointAt(0));
    const rightPoints = Array.from(right, character => character.codePointAt(0));
    const length = Math.min(leftPoints.length, rightPoints.length);
    for (let index = 0; index < length; index += 1) {
        if (leftPoints[index] < rightPoints[index]) return -1;
        if (leftPoints[index] > rightPoints[index]) return 1;
    }
    return leftPoints.length - rightPoints.length;
}


function float64Hex(value, path) {
    if (!Number.isFinite(value)) {
        throw narrowEditError(
            "narrow_edit_payload_invalid",
            `Operation payload numbers must be finite (at ${path}).`,
        );
    }
    const normalized = Object.is(value, -0) ? 0 : value;
    const bytes = new Uint8Array(8);
    new DataView(bytes.buffer).setFloat64(0, normalized, false);
    return [...bytes].map(byte => byte.toString(16).padStart(2, "0")).join("");
}


// path is diagnostic only (it names the exact offending field when this
// throws on a real, deeply nested workflow/prompt payload) - it never
// affects the canonical text or its hash, so existing hashes are unchanged.
export function canonicalTypedText(value, path = "$") {
    if (value === null) return "n;";
    if (typeof value === "boolean") return value ? "b1;" : "b0;";
    if (typeof value === "number") return `d${float64Hex(value, path)};`;
    if (typeof value === "string") {
        const encoded = new TextEncoder().encode(value);
        return `s${encoded.length}:${utf8Hex(value, path)};`;
    }
    if (Array.isArray(value)) {
        return `a${value.length}:[${
            value.map((item, index) => canonicalTypedText(item, `${path}[${index}]`)).join("")
        }];`;
    }
    if (value && typeof value === "object") {
        const prototype = Object.getPrototypeOf(value);
        if (prototype !== Object.prototype && prototype !== null) {
            throw narrowEditError(
                "narrow_edit_payload_invalid",
                `Operation payloads must use the exact JSON data model `
                + `(at ${path}: a non-plain object was found).`,
            );
        }
        const keys = Object.keys(value).sort(compareUnicodeScalars);
        return `o${keys.length}:{${keys.map(
            key => `${canonicalTypedText(key, `${path}.<key>`)}`
                + `${canonicalTypedText(value[key], `${path}.${key}`)}`,
        ).join("")}};`;
    }
    throw narrowEditError(
        "narrow_edit_payload_invalid",
        `Operation payloads must use the exact JSON data model `
        + `(at ${path}: found ${value === undefined ? "undefined" : typeof value}).`,
    );
}


export async function sha256Hex(value) {
    const bytes = new TextEncoder().encode(value);
    if (globalThis.crypto?.subtle) {
        const digest = await globalThis.crypto.subtle.digest("SHA-256", bytes);
        return [...new Uint8Array(digest)]
            .map(byte => byte.toString(16).padStart(2, "0"))
            .join("");
    }
    if (typeof process !== "undefined" && process.versions?.node) {
        const { createHash } = await import("node:crypto");
        return createHash("sha256").update(bytes).digest("hex");
    }
    throw narrowEditError(
        "sha256_unavailable",
        "SHA-256 is unavailable in this browser.",
    );
}


export async function canonicalNarrowOperationHash(tool, payload) {
    if (typeof tool !== "string" || !tool || tool.length > 128) {
        throw narrowEditError(
            "narrow_edit_tool_invalid",
            "The narrow mutation tool name is invalid.",
        );
    }
    if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
        throw narrowEditError(
            "narrow_edit_payload_invalid",
            "The narrow mutation payload must be an object.",
        );
    }
    return await sha256Hex(canonicalTypedText({
        schema: NARROW_EDIT_OPERATION_SCHEMA,
        tool,
        payload,
    }));
}


function validateRequestHash(requestHash) {
    if (typeof requestHash !== "string" || !SHA256_PATTERN.test(requestHash)) {
        throw narrowEditError(
            "narrow_edit_request_hash_invalid",
            "The narrow mutation request digest is invalid.",
        );
    }
    return requestHash;
}


function cloneJson(value, label) {
    // Validate before cloning so receipts cannot retain DOM objects, promises,
    // image elements, undefined fields, or other non-portable page state.
    canonicalTypedText(value);
    try {
        return structuredClone(value);
    } catch (error) {
        throw narrowEditError(
            "narrow_edit_receipt_invalid",
            `${label} must be detached JSON data.`,
        );
    }
}


/**
 * Bounded page-lifetime ledger. It stores hashes plus caller-sanitized receipts
 * and bindings, never raw execution payloads (including prompt strings).
 */
export class NarrowEditOperationLedger {
    constructor({ ttlMs = 30 * 60 * 1000, maxEntries = 256, now = () => Date.now() } = {}) {
        if (!Number.isFinite(ttlMs) || ttlMs <= 0) {
            throw new TypeError("ttlMs must be positive");
        }
        if (!Number.isInteger(maxEntries) || maxEntries < 1) {
            throw new TypeError("maxEntries must be a positive integer");
        }
        this.ttlMs = ttlMs;
        this.maxEntries = maxEntries;
        this.now = now;
        this.entries = new Map();
    }

    _key(tool, operationId) {
        if (typeof tool !== "string" || !tool || tool.length > 128) {
            throw narrowEditError(
                "narrow_edit_tool_invalid",
                "The narrow mutation tool name is invalid.",
            );
        }
        return `${tool.length}:${tool}:${validateNarrowOperationId(operationId)}`;
    }

    _prune() {
        const now = this.now();
        for (const [key, entry] of this.entries) {
            if (entry.expiresAt <= now) this.entries.delete(key);
        }
    }

    _makeRoom() {
        if (this.entries.size < this.maxEntries) return;
        const completed = [...this.entries.entries()]
            .filter(([, entry]) => entry.state === "completed")
            .sort((left, right) => left[1].createdAt - right[1].createdAt);
        if (completed.length > 0) {
            this.entries.delete(completed[0][0]);
            return;
        }
        throw narrowEditError(
            "narrow_edit_ledger_capacity",
            "Too many narrow mutations are still pending; wait for one to finish before retrying.",
        );
    }

    _matchingEntry(tool, operationId, requestHash) {
        const key = this._key(tool, operationId);
        const digest = validateRequestHash(requestHash);
        const entry = this.entries.get(key) || null;
        if (entry && entry.requestHash !== digest) {
            throw narrowEditError(
                "narrow_edit_idempotency_conflict",
                "operation_id was already used with different mutation arguments.",
            );
        }
        return { key, digest, entry };
    }

    async _attestedReplay(entry, attest) {
        if (typeof attest !== "function") {
            throw narrowEditError(
                "narrow_edit_attestation_required",
                "An exact state attestation is required before replaying mutation success.",
            );
        }
        const valid = await attest(
            cloneJson(entry.receipt, "Mutation receipt"),
            cloneJson(entry.binding, "Mutation binding"),
        );
        if (valid !== true) {
            throw narrowEditError(
                "narrow_edit_replay_state_changed",
                "The workflow state no longer matches the completed mutation; it was not run again.",
            );
        }
        return {
            ...cloneJson(entry.receipt, "Mutation receipt"),
            operation_id: entry.operationId,
            operation_request_hash: entry.requestHash,
            already_applied: true,
            queued: entry.receipt?.queued === true,
        };
    }

    /** Execute at most once, coalesce concurrent retries, and attest later replay. */
    async run({
        tool,
        operationId,
        requestHash,
        requestPayload,
        executionPayload,
        execute,
        receipt = value => value,
        binding = () => ({}),
        attest,
    }) {
        this._prune();
        const { key, digest, entry } = this._matchingEntry(
            tool,
            operationId,
            requestHash,
        );
        const requestHashPromise = requestPayload === undefined
            ? Promise.resolve(digest)
            : canonicalNarrowOperationHash(tool, requestPayload);
        const executionHashPromise = canonicalNarrowOperationHash(tool, executionPayload);
        if (entry) {
            const observedRequestHash = await requestHashPromise;
            if (observedRequestHash !== digest) {
                throw narrowEditError(
                    "narrow_edit_request_hash_mismatch",
                    "The canonical mutation arguments do not match operation_request_hash.",
                );
            }
            // requestHash is the canonical public mutation contract. The
            // backend may re-resolve a post-commit graph after losing a reply,
            // so its derived transactional payload is allowed to differ and is
            // never executed when the public operation already exists.
            if (entry.state === "completed") {
                return await this._attestedReplay(entry, attest);
            }
            if (entry.promise) {
                const firstResult = await entry.promise;
                return {
                    ...cloneJson(firstResult, "Mutation receipt"),
                    operation_id: operationId,
                    operation_request_hash: digest,
                    already_applied: true,
                    queued: firstResult?.queued === true,
                };
            }
            throw narrowEditError(
                "narrow_edit_outcome_pending",
                "The prior mutation outcome is unknown and cannot be executed twice.",
            );
        }
        if (typeof execute !== "function") {
            throw new TypeError("execute must be a function");
        }
        this._makeRoom();
        const createdAt = this.now();
        const created = {
            tool,
            operationId,
            requestHash: digest,
            executionHash: null,
            executionHashPromise,
            state: "pending",
            createdAt,
            expiresAt: createdAt + this.ttlMs,
            promise: null,
            receipt: null,
            binding: null,
        };
        // Reserve synchronously before the digest promise yields. Concurrent
        // identical calls will observe this entry and join its one execution.
        this.entries.set(key, created);
        created.promise = (async () => {
            try {
                const observedRequestHash = await requestHashPromise;
                if (observedRequestHash !== digest) {
                    throw narrowEditError(
                        "narrow_edit_request_hash_mismatch",
                        "The canonical mutation arguments do not match operation_request_hash.",
                    );
                }
                created.executionHash = await executionHashPromise;
            } catch (error) {
                // Canonicalization happens before execute(), so this is a known
                // pre-mutation failure and needs no tombstone.
                this.entries.delete(key);
                throw error;
            }
            const result = await execute();
            const safeReceipt = cloneJson(await receipt(result), "Mutation receipt");
            const safeBinding = cloneJson(await binding(result), "Mutation binding");
            created.receipt = safeReceipt;
            created.binding = safeBinding;
            created.state = "completed";
            created.expiresAt = this.now() + this.ttlMs;
            created.promise = null;
            return safeReceipt;
        })();
        try {
            const firstReceipt = await created.promise;
            return {
                ...cloneJson(firstReceipt, "Mutation receipt"),
                operation_id: operationId,
                operation_request_hash: digest,
                already_applied: false,
                queued: firstReceipt?.queued === true,
            };
        } catch (error) {
            // Keep an unknown/pending tombstone. Retrying with the same ID must
            // recover or fail closed, never execute the mutation a second time.
            created.promise = null;
            throw error;
        }
    }

    /** Recover a prior completed result without executing any mutation. */
    async recover({ tool, operationId, requestHash, attest }) {
        this._prune();
        const { entry } = this._matchingEntry(tool, operationId, requestHash);
        if (!entry) {
            throw narrowEditError(
                "narrow_edit_operation_not_found",
                "No bounded browser receipt exists for this operation; it was not run again.",
            );
        }
        if (entry.state === "completed") {
            return await this._attestedReplay(entry, attest);
        }
        if (entry.promise) {
            const result = await entry.promise;
            return {
                ...cloneJson(result, "Mutation receipt"),
                operation_id: operationId,
                operation_request_hash: requestHash,
                already_applied: true,
                queued: result?.queued === true,
            };
        }
        throw narrowEditError(
            "narrow_edit_outcome_pending",
            "The prior mutation outcome is unknown and cannot be executed twice.",
        );
    }

    /** Release only a caller-proven pre-mutation failure. */
    discardPending({ tool, operationId, requestHash }) {
        this._prune();
        const { key, entry } = this._matchingEntry(tool, operationId, requestHash);
        if (!entry || entry.state !== "pending" || entry.promise) return false;
        this.entries.delete(key);
        return true;
    }

    clear() {
        this.entries.clear();
    }

    get size() {
        this._prune();
        return this.entries.size;
    }
}


// Module lifetime matches the ComfyUI page and survives executor reconnects.
export const narrowEditOperationLedger = new NarrowEditOperationLedger();
